# Exploratory Data Analysis (EDA)
***
## Table of Contents
1. [Importing Data](#importing_data)
2. [Data Cleaning](#data_cleaning)
3. [Data Transformation](#data_transformation)
***



### UK Train Rides
Mock train ticket data for National Rail in the UK, from Jan to Apr 2024, including details on the type of ticket, the date & time for each journey, the departure & arrival stations, the ticket price, and more.
Raw data(railway.csv) extracted from [Maven Analytics - Playground](https://mavenanalytics.io/data-playground)

**Transaction ID**

- Unique identifier for an individual train ticket purchase

**Date of Purchase**

- Date the ticket was purchased

**Time of Purchase**

- Time the ticket was purchased


**Purchase Type**

- Whether the ticket was purchased online or directly at a train station

**Payment Method**

- Payment method used to purchase the ticket (Contactles Credit Card or Debit Card)

**Railcard**

- Whether the passenger is a National Railcard holder (Adult Senior or Disabled) or not (None). Railcard holders get 1/3 off their ticket purchases.

**Ticket Class**

- Seat class for the ticket (Standard or First)

**Ticket Type**

- When you bought or can use the ticket. Advance tickets are 1/2 off and must be purchased at least a day prior to departure. Off-Peak tickets are 1/4 off and must be used outside of peak hours (weekdays between 6-8am and 4-6pm). Anytime tickets are full price and can be bought and used at any time during the day.

**Price**

- Final cost of the ticket

**Departure Station**

- Station to board the train

***


In [715]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

## 1. Importing Data <a class="anchor" id="importing_data"></a>

In [716]:
df = pd.read_csv('railway.csv')
df.head()

Transaction ID Date of Purchase Time of Purchase Purchase Type  \
0  da8a6ba8-b3dc-4677-b176       2023-12-08         12:41:11        Online   
1  b0cdd1b0-f214-4197-be53       2023-12-16         11:23:01       Station   
2  f3ba7a96-f713-40d9-9629       2023-12-19         19:51:27        Online   
3  b2471f11-4fe7-4c87-8ab4       2023-12-20         23:00:36       Station   
4  2be00b45-0762-485e-a7a3       2023-12-27         18:22:56        Online   

  Payment Method Railcard Ticket Class Ticket Type  Price  \
0    Contactless    Adult     Standard     Advance     43   
1    Credit Card    Adult     Standard     Advance     23   
2    Credit Card      NaN     Standard     Advance      3   
3    Credit Card      NaN     Standard     Advance     13   
4    Contactless      NaN     Standard     Advance     76   

       Departure Station    Arrival Destination Date of Journey  \
0      London Paddington  Liverpool Lime Street      2024-01-01   
1     London Kings Cross                   York      2024-01-01   
2  Liverpool Lime Street  Manchester Piccadilly      2024-01-02   
3      London Paddington                Reading      2024-01-01   
4  Liverpool Lime Street          London Euston      2024-01-01   

  Departure Time Arrival Time Actual Arrival Time Journey Status  \
0       11:00:00     13:30:00            13:30:00        On Time   
1       09:45:00     11:35:00            11:40:00        Delayed   
2       18:15:00     18:45:00            18:45:00        On Time   
3       21:30:00     22:30:00            22:30:00        On Time   
4       16:45:00     19:00:00            19:00:00        On Time   

  Reason for Delay Refund Request  
0              NaN             No  
1   Signal Failure             No  
2              NaN             No  
3              NaN             No  
4              NaN             No

***

## 2. Data Cleaning <a class="anchor" id="data_cleaning"></a>
- Verify data types and formats.
- Handle missing values and duplicates.
- Drop unnecessary columns if required for further analysis / modelling.

In [717]:
df.dtypes

Transaction ID         object
Date of Purchase       object
Time of Purchase       object
Purchase Type          object
Payment Method         object
Railcard               object
Ticket Class           object
Ticket Type            object
Price                   int64
Departure Station      object
Arrival Destination    object
Date of Journey        object
Departure Time         object
Arrival Time           object
Actual Arrival Time    object
Journey Status         object
Reason for Delay       object
Refund Request         object
dtype: object

In [718]:
df['Purchase Type'].value_counts()

Purchase Type
Online     18521
Station    13132
Name: count, dtype: int64

In [719]:
df['Payment Method'].value_counts()

Payment Method
Credit Card    19136
Contactless    10834
Debit Card      1683
Name: count, dtype: int64

In [720]:
df['Railcard'].value_counts()

Railcard
Adult       4846
Disabled    3089
Senior      2800
Name: count, dtype: int64

In [721]:
df['Ticket Class'].value_counts()

Ticket Class
Standard       28595
First Class     3058
Name: count, dtype: int64

In [722]:
df['Ticket Type'].value_counts()

Ticket Type
Advance     17561
Off-Peak     8752
Anytime      5340
Name: count, dtype: int64

In [723]:
df['Journey Status'].value_counts()

Journey Status
On Time      27481
Delayed       2292
Cancelled     1880
Name: count, dtype: int64

In [724]:
df['Reason for Delay'].value_counts()

Reason for Delay
Weather               995
Technical Issue       707
Signal Failure        523
Signal failure        447
Staffing              410
Staff Shortage        399
Weather Conditions    377
Traffic               314
Name: count, dtype: int64

In [725]:
df['Refund Request'].value_counts()

Refund Request
No     30535
Yes     1118
Name: count, dtype: int64

`df.dtypes` shows that most of the features are stored as the `object` datatype. Let's convert them into their appropriate datatypes. **Time of Purchase**, **Departure Time**, **Arrival Time** and **Actual Arrival Time** will be left as `object` for now.
- **Transaction ID** -> String
- **Date of Purchase** -> Datetime (only YYYY-MM-DD)
- **Purchase Type** -> Categorical (nominal)
- **Payment Method** -> Categorical (nominal)
- **Railcard** -> Categorical (nominal)
- **Ticket Class** -> Categorical (nominal)
- **Ticket Type** -> Categorical (nominal)
- **Departure Station** -> String
- **Arrival Destination** -> String
- **Date of Journey** -> Datetime (only YYYY-MM-DD)
- **Journey Status** -> Categorical (nominal)
- **Reason for Delay** -> Categorical (nominal)
- **Refund Request** -> Boolean (True: Yes, False: No)

In [726]:
df['Transaction ID'] = df['Transaction ID'].astype('string')
df['Date of Purchase'] = pd.to_datetime(df['Date of Purchase'])
df['Purchase Type'] = df['Purchase Type'].astype('category')
df['Payment Method'] = df['Payment Method'].astype('category')
df['Railcard'] = df['Railcard'].astype('category')
df['Ticket Class'] = df['Ticket Class'].astype('category')
df['Ticket Type'] = df['Ticket Type'].astype('category')
df['Departure Station'] = df['Departure Station'].astype('string')
df['Arrival Destination'] = df['Arrival Destination'].astype('string')
df['Date of Journey'] = pd.to_datetime(df['Date of Journey'])
df['Journey Status'] = df['Journey Status'].astype('category')
df['Reason for Delay'] = df['Reason for Delay'].astype('category')
df['Refund Request'] = df['Refund Request'].map({'Yes':True, 'No':False})

In [727]:
df.dtypes

Transaction ID         string[python]
Date of Purchase       datetime64[ns]
Time of Purchase               object
Purchase Type                category
Payment Method               category
Railcard                     category
Ticket Class                 category
Ticket Type                  category
Price                           int64
Departure Station      string[python]
Arrival Destination    string[python]
Date of Journey        datetime64[ns]
Departure Time                 object
Arrival Time                   object
Actual Arrival Time            object
Journey Status               category
Reason for Delay             category
Refund Request                   bool
dtype: object

In [728]:
df.isnull().sum()

Transaction ID             0
Date of Purchase           0
Time of Purchase           0
Purchase Type              0
Payment Method             0
Railcard               20918
Ticket Class               0
Ticket Type                0
Price                      0
Departure Station          0
Arrival Destination        0
Date of Journey            0
Departure Time             0
Arrival Time               0
Actual Arrival Time     1880
Journey Status             0
Reason for Delay       27481
Refund Request             0
dtype: int64

20918 , 1880, 27481 null values were detected in the `Railcard`, `Actual Arrival Time`, `Reason for Delay` columns respectively.

In [729]:
df[df['Actual Arrival Time'].isnull()]['Journey Status'].value_counts()

Journey Status
Cancelled    1880
Delayed         0
On Time         0
Name: count, dtype: int64

In [730]:
df[df['Reason for Delay'].isnull()]['Journey Status'].value_counts()

Journey Status
On Time      27481
Cancelled        0
Delayed          0
Name: count, dtype: int64

For the `Railcard` column, let's assume there are 4 types of card - Youth, Adult, Disabled and Senior. Null values will be replaced with `Youth`.

In [731]:
df['Railcard'] = df['Railcard'].cat.add_categories('Youth')
df['Railcard'] = df['Railcard'].fillna('Youth')
df['Railcard'].value_counts()

Railcard
Youth       20918
Adult        4846
Disabled     3089
Senior       2800
Name: count, dtype: int64

All null values in the `Actual Arrival Time` column are due to cancellations. These null values will be replaced with `0`.

In [732]:
df['Actual Arrival Time'] = df['Actual Arrival Time'].fillna(0)
df['Actual Arrival Time'].isnull().sum()

np.int64(0)

 Similarly, null values in the `Reason for Delay` column occur only when the `Journey Status` is `On Time`. These will be replaced with `Not Delayed`.

In [733]:
df['Reason for Delay'] = df['Reason for Delay'].cat.add_categories('Not Delayed')
df['Reason for Delay'] = df['Reason for Delay'].fillna('Not Delayed')
df['Reason for Delay'].isnull().sum()

np.int64(0)

In [734]:
df.isnull().sum()

Transaction ID         0
Date of Purchase       0
Time of Purchase       0
Purchase Type          0
Payment Method         0
Railcard               0
Ticket Class           0
Ticket Type            0
Price                  0
Departure Station      0
Arrival Destination    0
Date of Journey        0
Departure Time         0
Arrival Time           0
Actual Arrival Time    0
Journey Status         0
Reason for Delay       0
Refund Request         0
dtype: int64

## 3. Feature Engineering <a class="anchor" id="feature_engineering"></a>
- Normalise or standardise numerical variables if necessary for further analysis / modelling.
- Encode categorical variables if necessary.
- Create new features from existing ones.

In this analysis, there is no need of normalising, standarsing numerical variables or encoding categorical variables as creating machine learning models is not a part of this project. Instead, a couple of new features will be created: `Days Before Depature`, `Travel Time`, and `Delayed Time`.


**Days Before Departure**
- Number of days in advance passengers purchased their tickets before departure, calculated as the difference between `Date of Purchase` and `Date of Journey`.

**Travel Time**
- Total travel time in minutes, calculated as the difference between `Departure Time` and `Actual Arrival Time`.

**Delayed Time**

- Delay duration in minutes for delayed trains, calculated as the difference between `Arrival Time` and `Actual Arrival Time`.

In [749]:
# Create `Days Before Departure`
df['Days Before Departure'] = (df['Date of Journey'] - df['Date of Purchase']).dt.days

# Create `Travel Time`
# Convert to datetime format -> Calculate in minutes -> Handle null values and convert to int
departure_time_dt = pd.to_datetime(df['Departure Time'], format='%H:%M:%S')
actual_arrival_time_dt = pd.to_datetime(df['Actual Arrival Time'], format='%H:%M:%S', errors='coerce') # errors='coerce' handles invalid or null values
df['Travel Time'] = np.where(~df['Actual Arrival Time'].isnull(), (actual_arrival_time_dt - departure_time_dt).dt.total_seconds()/60, 0)
df['Travel Time'] = df['Travel Time'].fillna(0).astype(int)

# Create `Delayed Time`
arrival_time_dt = pd.to_datetime(df['Arrival Time'], format='%H:%M:%S')
df['Delayed Time'] = np.where(df['Journey Status'] == 'Delayed', (actual_arrival_time_dt - arrival_time_dt).dt.total_seconds()/60, 0)
df['Delayed Time'] = df['Delayed Time'].fillna(0).astype(int)

In [751]:
df.head()

Transaction ID Date of Purchase Time of Purchase Purchase Type  \
0  da8a6ba8-b3dc-4677-b176       2023-12-08         12:41:11        Online   
1  b0cdd1b0-f214-4197-be53       2023-12-16         11:23:01       Station   
2  f3ba7a96-f713-40d9-9629       2023-12-19         19:51:27        Online   
3  b2471f11-4fe7-4c87-8ab4       2023-12-20         23:00:36       Station   
4  2be00b45-0762-485e-a7a3       2023-12-27         18:22:56        Online   

  Payment Method Railcard Ticket Class Ticket Type  Price  \
0    Contactless    Adult     Standard     Advance     43   
1    Credit Card    Adult     Standard     Advance     23   
2    Credit Card    Youth     Standard     Advance      3   
3    Credit Card    Youth     Standard     Advance     13   
4    Contactless    Youth     Standard     Advance     76   

       Departure Station  ... Date of Journey Departure Time Arrival Time  \
0      London Paddington  ...      2024-01-01       11:00:00     13:30:00   
1     London Kings Cross  ...      2024-01-01       09:45:00     11:35:00   
2  Liverpool Lime Street  ...      2024-01-02       18:15:00     18:45:00   
3      London Paddington  ...      2024-01-01       21:30:00     22:30:00   
4  Liverpool Lime Street  ...      2024-01-01       16:45:00     19:00:00   

  Actual Arrival Time Journey Status Reason for Delay Refund Request  \
0            13:30:00        On Time      Not Delayed          False   
1            11:40:00        Delayed   Signal Failure          False   
2            18:45:00        On Time      Not Delayed          False   
3            22:30:00        On Time      Not Delayed          False   
4            19:00:00        On Time      Not Delayed          False   

   Days Before Departure  Travel Time  Delayed Time  
0                     24          150             0  
1                     16          115             5  
2                     14           30             0  
3                     12           60             0  
4                      5          135             0  

[5 rows x 21 columns]

## 4. Feature Engineering <a class="anchor" id="feature_engineering"></a>